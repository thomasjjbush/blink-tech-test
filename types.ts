import { AnyAction } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';

export type Action = ThunkAction<Promise<AnyAction | void>, Store, unknown, AnyAction>;

export type Dispatch = ThunkDispatch<Store, void, AnyAction>;

export interface Conversation {
    id: string;
    last_updated: string;
    name: string;
}

export enum ConversationActions {
    LOADED = 'conversation/LOADED',
    UPDATED = 'conversation/UPDATED',
}

export interface ConversationsProps {
    conversation: string;
    setConversation: (conversation: string) => void;
}

export enum Endpoints {
    CONVERSATIONS = 'conversations',
    MESSAGES = 'messages',
}

export interface ErrorState {
    code?: number;
    fatal?: boolean;
    message?: string;
}

export enum ErrorActions {
    'FATAL' = 'error/FATAL',
}

export interface Message {
    conversation: string;
    edited?: true;
    id: string;
    isUser?: true;
    last_updated: string;
    text: string;
}

export enum MessageActions {
    EDIT = 'messages/EDIT',
    LOADED = 'messages/LOADED',
    SEND = 'messages/SEND',
}

export interface MessageBubbleProps extends Message {
    onClick: (id: string, text: string) => void;
}

export interface MessageState {
    [key: string]: Message[];
}

export interface MessagesProps {
    conversation: string;
}

export interface Options {
    [key: string]: string | number;
    conversation?: string;
    _sort?: string;
}

export interface Store {
    conversations: Conversation[];
    error: ErrorState;
    messages: MessageState;
}

export type StyledProps<Props = unknown> = { theme: Theme } & Props;

export interface TextAreaProps {
    setValue: (value: string) => void;
    value: string;
}

export interface Theme {
    brightBlue: string;
    darkBlue: string;
    lightBlue: string;
}
