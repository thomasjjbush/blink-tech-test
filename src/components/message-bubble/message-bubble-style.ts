import styled from 'styled-components';
import { StyledProps } from '../../../types';

export const Bubble = styled.div<StyledProps<{ isUser: boolean }>>`
    align-self: ${({ isUser }) => (isUser ? 'flex-end' : 'auto')};
    background-color: ${({ isUser, theme }) => (isUser ? theme.brightBlue : theme.lightBlue)};
    border-radius: 20px;
    color: ${({ isUser }) => (isUser ? 'white' : 'auto')};
    display: inline;
    margin-bottom: 20px;
    max-width: 350px;
    padding: 20px;
    position: relative;

    &:after {
        border-left: 10px solid transparent;
        border-right: 10px solid transparent;
        border-top: 10px solid ${({ isUser, theme }) => (isUser ? theme.brightBlue : theme.lightBlue)};
        content: '';
        ${({ isUser }) => (isUser ? 'right: 20px;' : 'left: 20px')};
        position: absolute;
        top: 100%;
        width: 0;
    }
`;

export const Button = styled.button`
    appearance: none;
    background-color: transparent;
    border: none;
    color: inherit;
    cursor: pointer;
    margin-right: 5px;
    padding: 0;
    text-decoration: underline;
`;

export const Text = styled.p`
    margin: 0 0 10px;
`;
