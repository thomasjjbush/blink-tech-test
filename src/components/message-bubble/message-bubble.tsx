import React, { FC, ReactElement } from 'react';
import { DateAndTime } from '../date-and-time/date-and-time';
import { MessageBubbleProps as Props } from '../../../types';
import * as Styled from './message-bubble-style';

export const MessageBubble: FC<Props> = ({ id, isUser, last_updated, onClick, text }: Props): ReactElement => (
    <Styled.Bubble isUser={isUser}>
        <Styled.Text>{text}</Styled.Text>
        {isUser && (
            <Styled.Button data-test-id="edit-button" onClick={() => onClick(id, text)} type="button">
                Edit
            </Styled.Button>
        )}
        <DateAndTime>{last_updated}</DateAndTime>
    </Styled.Bubble>
);
