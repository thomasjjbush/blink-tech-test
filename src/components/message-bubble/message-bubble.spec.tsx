import React from 'react';
import { shallow } from 'enzyme';
import { MessageBubbleProps as Props } from '../../../types';
import { MessageBubble } from './message-bubble';

describe('MessageBubble', () => {
    let props: Props;

    beforeEach(() => {
        props = {
            conversation: 'convo',
            id: 'id',
            last_updated: 'a long time ago, in a galaxy far away',
            onClick: jest.fn(),
            text: 'text',
        };
    });

    it('should match snapshot for non user message', () => {
        expect(shallow(<MessageBubble {...props} />)).toMatchSnapshot();
    });

    it('should match snapshot for user message (apply correct conditional styles and render edit button)', () => {
        expect(shallow(<MessageBubble {...props} isUser />)).toMatchSnapshot();
    });

    it('should invoke onClick cb with correct arguments on edit button click', () => {
        const wrapper = shallow(<MessageBubble {...props} isUser />);
        wrapper.find({ 'data-test-id': 'edit-button' }).simulate('click');

        expect(props.onClick).toHaveBeenCalledTimes(1);
        expect(props.onClick).toHaveBeenCalledWith('id', 'text');
    });
});
