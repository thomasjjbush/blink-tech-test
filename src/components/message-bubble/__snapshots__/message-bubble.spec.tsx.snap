// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageBubble should match snapshot for non user message 1`] = `
<styled.div
  styles="
    align-self: auto;
    background-color: theme.lightBlue;
    border-radius: 20px;
    color: auto;
    display: inline;
    margin-bottom: 20px;
    max-width: 350px;
    padding: 20px;
    position: relative;

    &:after {
        border-left: 10px solid transparent;
        border-right: 10px solid transparent;
        border-top: 10px solid theme.lightBlue;
        content: '';
        left: 20px;
        position: absolute;
        top: 100%;
        width: 0;
    }
"
>
  <styled.p
    styles="
    margin: 0 0 10px;
    white-space: pre-wrap;
"
  >
    text
  </styled.p>
  <DateAndTime>
    a long time ago, in a galaxy far away
  </DateAndTime>
</styled.div>
`;

exports[`MessageBubble should match snapshot for user message (apply correct conditional styles and render edit button) 1`] = `
<styled.div
  isUser={true}
  styles="
    align-self: flex-end;
    background-color: theme.brightBlue;
    border-radius: 20px;
    color: white;
    display: inline;
    margin-bottom: 20px;
    max-width: 350px;
    padding: 20px;
    position: relative;

    &:after {
        border-left: 10px solid transparent;
        border-right: 10px solid transparent;
        border-top: 10px solid theme.brightBlue;
        content: '';
        right: 20px;;
        position: absolute;
        top: 100%;
        width: 0;
    }
"
>
  <styled.p
    styles="
    margin: 0 0 10px;
    white-space: pre-wrap;
"
  >
    text
  </styled.p>
  <styled.button
    data-test-id="edit-button"
    onClick={[Function]}
    styles="
    appearance: none;
    background-color: transparent;
    border: none;
    color: inherit;
    cursor: pointer;
    margin-right: 5px;
    padding: 0;
    text-decoration: underline;
"
    type="button"
  >
    Edit
  </styled.button>
  <DateAndTime>
    a long time ago, in a galaxy far away
  </DateAndTime>
</styled.div>
`;
